client/verta/verta/pipeline/_pipeline_graph.py

```diff
@@ -0,0 +1,146 @@
+# -*- coding: utf-8 -*-
+
+from typing import Any, Dict, List, Set, Tuple, Union
+
+from verta._internal_utils._utils import Configuration, Connection
+from ._pipeline_step import PipelineStep
+
+
+class PipelineGraph:
+    """Object representing a collection of PipelineSteps to be run as a single
+    inference pipeline.
+
+    Parameters
+    ----------
+    steps : list, set, or tuple of :class:`~verta.pipeline.PipelineStep`
+        Set of all possible steps of the pipeline. Ordering of steps in the pipeline
+        itself is determined by the predecessors provided to each step.
+
+    Attributes
+    ----------
+    steps: set of :class:`~verta.deployment.PipelineStep`
+        Set of PipelineSteps comprising all possible steps in this PiplineGraph.
+    """
+
+    def __init__(self, steps: Set[PipelineStep]):
+        self._steps = self._validate_steps(steps)
+        for step in self._steps:
+            step._validate_predecessors(step.predecessors)
+        # throws an exception if any step's predecessors attr has been inappropriately mutated.
+
+    def __repr__(self) -> str:
+        return f"PipelineGraph steps:\n{self._format_steps()}"
+
+    def _format_steps(self) -> str:
+        """Format steps for improved readability in __repr__() function."""
+        return "\n".join([repr(s) for s in self._steps])
```

Comment:
Ideally, we ought to indent the steps one level in, similar to https://github.com/VertaAI/modeldb/pull/4020#discussion_r1306144906, but we can address that later.

![image](https://github.com/VertaAI/modeldb/assets/96442646/dc52b772-6a0d-4316-839b-6e882dc3e6e1)

Reply:
Created a new [ticket](https://vertaai.atlassian.net/browse/VRD-1103) to address this